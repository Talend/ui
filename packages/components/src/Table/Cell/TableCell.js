import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

function isClickable(props) {
	return props.onClick || props.onDoubleClick || props.onKeyPress;
}

function renderCell(data, className) {
	return <div className={classnames('tc-table-cell', className)}>{data}</div>;
}

function renderClickableCell(data, className, onClick, onDoubleClick, onKeyPress) {
	return (
		/* eslint-disable jsx-a11y/no-static-element-interactions */
		<div
			className={classnames('tc-table-clickable-cell', className)}
			onClick={onClick}
			onDoubleClick={onDoubleClick}
			onKeyPress={onKeyPress}
			role="button"
			tabIndex="-1"
		>
			{data}
		</div>
		/* eslint-enable jsx-a11y/no-static-element-interactions */
	);
}

/**
 * This component displays a cell of a table as a clickable element. The data must be a string.
 * Callback methods (onClick, onDoubleClick and onKeyPress) must be passed via props.
 */
export default class TableCell extends Component {
	constructor(props) {
		super(props);
		this.onClick = this.onClick.bind(this);
		this.onDoubleClick = this.onDoubleClick.bind(this);
		this.onKeyPress = this.onKeyPress.bind(this);
	}

	onClick(ev) {
		if (this.props.onClick) {
			this.props.onClick(this.props.element, ev);
		}
	}

	onDoubleClick(ev) {
		if (this.props.onDoubleClick) {
			this.props.onDoubleClick(this.props.element, ev);
		}
	}

	onKeyPress(ev) {
		if (this.props.onKeyPress) {
			this.props.onKeyPress(this.props.element, ev);
		}
	}

	render() {
		const { data, className } = this.props;
		if (isClickable(this.props)) {
			return renderClickableCell(
				data,
				className,
				this.onClick,
				this.onDoubleClick,
				this.onKeyPress,
			);
		}
		return renderCell(data, className);
	}
}

TableCell.propTypes = {
	element: PropTypes.object,
	data: PropTypes.string,
	className: PropTypes.string,
	onClick: PropTypes.func,
	onDoubleClick: PropTypes.func,
	onKeyPress: PropTypes.func,
};
