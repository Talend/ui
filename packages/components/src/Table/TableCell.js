import React from 'react';
import PropTypes from 'prop-types';

/**
 * This component displays a cell of a table as a simple label. The data must be a string.
 * It uses a div element for layout purpose (text alignment in a cell).
 */
export default function TableCell({ data, className }) {
	return (
		<div
			className={`tc-table-cell ${className}`}
		>
			{data}
		</div>
	);
}

TableCell.propTypes = {
	data: PropTypes.string,
	className: PropTypes.string,
};
