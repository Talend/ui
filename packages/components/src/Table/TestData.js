import StringFilterComponent from './Filters/StringFilterComponent';
import ToggleFilterComponent from './Filters/ToggleFilterComponent';
// This module defines data, objects and functions used by the unit tests.

export const element1 = {
	id: 'elem_1',
	name: 'element 1',
	type: 'string',
	desc: 'Description of element 1',
};

export const element2 = {
	id: 'elem_2',
	name: 'element 2',
	type: 'string',
	desc: 'This element is a string',
};

export const firstname = {
	id: 'elem_firstname',
	name: 'Firstname',
	type: 'string',
	desc: 'The firstname of the customer (optional)',
	mandatory: false,
};

export const lastname = {
	id: 'elem_lastname',
	name: 'Lastname',
	type: 'string',
	desc: 'The lastname of the customer (mandatory)',
	mandatory: true,
};

export const birthday = {
	id: 'elem_birthday',
	name: 'Birthday',
	type: 'date',
	desc: 'The birthday of the customer (optional)',
	mandatory: false,
};

export const address = {
	id: 'elem_address',
	name: 'Address',
	type: 'address',
	desc: 'The address of the customer (mandatory)',
	mandatory: true,
};

export const Columns = {
	NAME: {
		key: 'name',
		label: 'Name',
	},
	TYPE: {
		key: 'type',
		label: 'Type',
	},
	DESC: {
		key: 'desc',
		label: 'Description',
	},
	MANDATORY: {
		key: 'mandatory',
		label: '',
	},
};

export const columns1 = [Columns.NAME, Columns.TYPE, Columns.DESC];

function getColumn(col) {
	return {
		key: col.key,
		label: col.label,
		headClassName: `classname-of-head-${col.key}`,
		cellClassName: `classname-of-row-data-${col.key}`,
	};
}

export const columns2 = [getColumn(Columns.NAME), getColumn(Columns.TYPE), getColumn(Columns.DESC)];

export const classnames = {
	titleBar: 'classname-of-title-bar',
	title: 'classname-of-title',
	filtersBar: 'classname-of-filters-bar',
	table: 'classname-of-table',
	header: 'classname-of-table-header',
	body: 'classname-of-table-body',
	row: 'classname-of-row',
	rows: {
		elem_1: 'classname-of-row-1',
		elem_2: 'classname-of-row-2',
	},
};

const nameFilterId = 'name-filter';
export const nameFilter = {
	id: nameFilterId,
	active: false,
	match() {
		return true;
	},
	renderer: StringFilterComponent,
	rendererProps: {
		placeHolder: 'Filter...',
		dockable: true,
		navbar: true,
	},
	className: nameFilterId,
};

const mandatoryFieldFilterId = 'mandatory-field-filter';
export const mandatoryFieldFilter = {
	id: mandatoryFieldFilterId,
	active: false,
	match() {
		return true;
	},
	renderer: ToggleFilterComponent,
	rendererProps: {
		label: 'Show Mandatory Fields (*) Only',
	},
	className: mandatoryFieldFilterId,
};
