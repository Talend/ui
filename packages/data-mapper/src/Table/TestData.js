import StringFilterComponent from './Filters/StringFilterComponent';
import ToggleFilterComponent from './Filters/ToggleFilterComponent';
// This module defines data, objects and functions used by the unit tests.

export const element1 = {
	id: 'elem_1',
	name: 'element 1',
	type: 'string',
	desc: 'Description of element 1',
};

export const element2 = {
	id: 'elem_2',
	name: 'element 2',
	type: 'string',
	desc: 'This element is a string',
};

export const firstname = {
	id: 'elem_firstname',
	name: 'Firstname',
	type: 'string',
	desc: 'The firstname of the customer (optional)',
	mandatory: false,
};

export const lastname = {
	id: 'elem_lastname',
	name: 'Lastname',
	type: 'string',
	desc: 'The lastname of the customer (mandatory)',
	mandatory: true,
};

export const birthday = {
	id: 'elem_birthday',
	name: 'Birthday',
	type: 'date',
	desc: 'The birthday of the customer (optional)',
	mandatory: false,
};

export const address = {
	id: 'elem_address',
	name: 'Address',
	type: 'address',
	desc: 'The address of the customer (mandatory)',
	mandatory: true,
};

export const Columns = {
	NAME: {
		id: 'my-table-name',
		key: 'name',
		label: 'Name',
	},
	TYPE: {
		id: 'my-table-type',
		key: 'type',
		label: 'Type',
	},
	DESC: {
		id: 'my-table-desc',
		key: 'desc',
		label: 'Description',
	},
	MANDATORY: {
		id: 'my-table-mandatory',
		key: 'mandatory',
		label: '',
	},
};

export const columns1 = [Columns.NAME, Columns.TYPE, Columns.DESC];

export const rowsClassName = {
	elem_1: 'classname-of-row-1',
	elem_2: 'classname-of-row-2',
};

export const nameFilterId = 'name-filter';

export function getNameFilter(id, match) {
	return {
		id,
		active: false,
		params: {
			value: null,
			docked: true,
		},
		match,
		renderer: StringFilterComponent,
		rendererProps: {
			placeHolder: 'Filter...',
			dockable: true,
			navbar: true,
		},
	};
}

export const mandatoryFieldFilterId = 'mandatory-field-filter';

export function getMandatoryFieldFilter(id, match) {
	return {
		id,
		active: false,
		match,
		renderer: ToggleFilterComponent,
		rendererProps: {
			label: 'Show Mandatory Fields (*) Only',
		},
	};
}

export const SortDirection = {
	NONE: 'none',
	ASCENDING: 'ascending',
	DESCENDING: 'descending',
};

export const sorterIcons = {
	none: 'none-icon',
	ascending: 'ascending-icon',
	descending: 'descending-icon',
};

export function addSortExtraProps(column) {
	return {
		...column,
		headExtraProps: {
			iconPosition: 'right',
			link: true,
		},
	};
}

export function getSorter() {
	return {
		direction: SortDirection.NONE,
		icons: sorterIcons,
	};
}
