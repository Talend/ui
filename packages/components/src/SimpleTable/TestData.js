import Cell from './Cell';
import Header from './Header';

export const element1 = {
	id: 'elem_1',
	name: 'element 1',
	type: 'string',
	desc: 'Description of element 1',
};

export const element2 = {
	id: 'elem_2',
	name: 'element 2',
	type: 'string',
	desc: 'This element is a string',
};

export const KEYS = {
	NAME: 'name',
	TYPE: 'type',
	DESC: 'desc',
};

export const columnKeys = [KEYS.NAME, KEYS.TYPE, KEYS.DESC];

export const classNameProvider = {
	getForTable() {
		return 'classname-of-simple-table';
	},
	getForHeader(columnKey) {
		return `classname-of-header-${columnKey}`;
	},
	getForRow(element) {
		return `classname-of-row-(${element.id})`;
	},
	getForRowData(element, columnKey) {
		return `classname-of-row-data-(${element.id},${columnKey})`;
	},
};

export const rowDataGetter = {
	getId(element) {
		return element.id;
	},
	getHeaderData(columnKey) {
		return columnKey;
	},
	getRowData(element, columnKey) {
		switch (columnKey) {
			case KEYS.NAME:
				return element.name;
			case KEYS.TYPE:
				return element.type;
			case KEYS.DESC:
				return element.desc;
			default:
				return `No data available for ${columnKey}`;
		}
	},
};

export const rowRenderer = {
	needRowUpdate(props) {
		return true;
	},
	getCellComponent(columnKey) {
		return Cell;
	},
	getExtraProps(columnKey) {
		return null;
	},
};

export const headerRenderer = {
	getHeaderComponent(columnKey) {
		return Header;
	},
	getExtraProps(columnKey) {
		return null;
	},
};
