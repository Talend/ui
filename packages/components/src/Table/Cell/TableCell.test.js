import React from 'react';
import renderer from 'react-test-renderer';
import TableCell from './TableCell';
import * as TestData from '../TestData';

/**
 * Render a single cell
 */
it('single-cell', () => {
  const element = TestData.element1;
	// create React tree
	const tree = renderer
		.create(
			<TableCell
				element={element}
				data={element.name}
        className="my-single-cell"
			/>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

/**
 * Render a clickable cell
 */
it('clickable-cell', () => {
  const element = TestData.element1;
	// create React tree
	const tree = renderer
		.create(
			<TableCell
				element={element}
				data={element.name}
        className="my-clickable-cell"
        onClick={jest.fn()}
				onDoubleClick={jest.fn()}
				onKeyPress={jest.fn()}
			/>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
