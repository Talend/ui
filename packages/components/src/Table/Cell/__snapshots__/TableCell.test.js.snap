// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clickable-cell 1`] = `
<div
  className="tc-table-clickable-cell my-clickable-cell"
  onClick={[Function]}
  onDoubleClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="-1"
>
  element 1
</div>
`;

exports[`single-cell 1`] = `
<div
  className="tc-table-cell my-single-cell"
>
  element 1
</div>
`;
